File: src/main/java/com/sotasan/decompiler/views/TreeView.java
package com.sotasan.decompiler.views;

import com.formdev.flatlaf.extras.components.FlatScrollPane;
import com.formdev.flatlaf.extras.components.FlatTree;
import com.sotasan.decompiler.controllers.TabsController;
import com.sotasan.decompiler.models.BaseModel;
import com.sotasan.decompiler.models.FileModel;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import javax.swing.*;
import javax.swing.tree.*;
import java.awt.*;
import java.awt.event.*;

@Getter
public class TreeView extends JPanel {

    private final FlatTree tree;
    private final FlatScrollPane scrollPane;

    public TreeView() {
        super(new BorderLayout());

        // TODO: only one root node
        tree = new FlatTree();
        tree.addKeyListener(new TreeKeyListener());
        tree.addMouseListener(new TreeMouseAdapter());
        tree.getSelectionModel().setSelectionMode(TreeSelectionModel.SINGLE_TREE_SELECTION);
        tree.setCellRenderer(new TreeCellRenderer());
        tree.setModel(new DefaultTreeModel(new DefaultMutableTreeNode()));
        tree.setRootVisible(false);
        tree.setShowsRootHandles(true);

        scrollPane = new FlatScrollPane();
        scrollPane.setBorder(BorderFactory.createEmptyBorder());
        scrollPane.setViewportView(tree);
        add(scrollPane);
    }

    private static void addTab(@NotNull InputEvent event) {
        TreePath path = ((JTree) event.getSource()).getSelectionPath();
        if (path == null)
            return;

        DefaultMutableTreeNode node = (DefaultMutableTreeNode) path.getLastPathComponent();
        if (node.getUserObject() == null)
            return;

        BaseModel model = (BaseModel) node.getUserObject();
        if (model instanceof FileModel)
            TabsController.getINSTANCE().addTab((FileModel) model);
    }

    private static class TreeCellRenderer extends DefaultTreeCellRenderer {

        @Override
        public Component getTreeCellRendererComponent(JTree tree, Object value, boolean selected, boolean expanded, boolean leaf, int row, boolean focused) {
            Component component = super.getTreeCellRendererComponent(tree, value, selected, expanded, leaf, row, focused);
            DefaultMutableTreeNode node = (DefaultMutableTreeNode) value;
            if (node.getUserObject() instanceof BaseModel) {
                BaseModel model = (BaseModel) node.getUserObject();
                setText(model.getName());
                setIcon(new ImageIcon(model.getIcon()));
            }
            return component;
        }

    }

    private static class TreeMouseAdapter extends MouseAdapter {

        @Override
        public void mousePressed(@NotNull MouseEvent event) {
            if (event.getClickCount() % 2 == 0)
                addTab(event);
        }

    }

    private static class TreeKeyListener implements KeyListener {

        @Override
        public void keyPressed(@NotNull KeyEvent keyEvent) {
            if (keyEvent.getExtendedKeyCode() == KeyEvent.VK_ENTER)
                addTab(keyEvent);
        }

        @Override public void keyTyped(KeyEvent keyEvent) {}
        @Override public void keyReleased(KeyEvent keyEvent) {}

    }

}

File: src/main/kotlin/moe/sota/decompiler/views/TreeView.kt
package moe.sota.decompiler.views

import com.formdev.flatlaf.extras.components.FlatScrollPane
import com.formdev.flatlaf.extras.components.FlatTree
import java.awt.BorderLayout
import java.awt.Component
import java.awt.event.InputEvent
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.BorderFactory
import javax.swing.ImageIcon
import javax.swing.JPanel
import javax.swing.JTree
import javax.swing.ToolTipManager
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeCellRenderer
import javax.swing.tree.DefaultTreeModel
import javax.swing.tree.TreeSelectionModel
import moe.sota.decompiler.controllers.TabsController
import moe.sota.decompiler.models.BaseModel
import moe.sota.decompiler.models.FileModel

class TreeView(private val tabsController: TabsController) : JPanel(BorderLayout()) {
    // TODO: only one root node
    val tree =
        FlatTree().apply {
            addKeyListener(TreeKeyListener(this@TreeView))
            addMouseListener(TreeMouseAdapter(this@TreeView))
            selectionModel.selectionMode = TreeSelectionModel.SINGLE_TREE_SELECTION
            cellRenderer = TreeCellRenderer()
            model = DefaultTreeModel(DefaultMutableTreeNode())
            isRootVisible = false
            showsRootHandles = true
        }

    init {
        ToolTipManager.sharedInstance().registerComponent(tree)

        val scrollPane =
            FlatScrollPane().apply {
                border = BorderFactory.createEmptyBorder()
                setViewportView(tree)
            }
        add(scrollPane)
    }

    fun addTab(event: InputEvent) {
        val path = (event.source as JTree).selectionPath ?: return
        val model = (path.lastPathComponent as DefaultMutableTreeNode).userObject
        if (model is FileModel) tabsController.addTab(model)
    }
}

private class TreeCellRenderer : DefaultTreeCellRenderer() {
    override fun getTreeCellRendererComponent(
        tree: JTree?,
        value: Any?,
        selected: Boolean,
        expanded: Boolean,
        leaf: Boolean,
        row: Int,
        focused: Boolean,
    ): Component? {
        val component =
            super.getTreeCellRendererComponent(tree, value, selected, expanded, leaf, row, focused)
        val model = (value as DefaultMutableTreeNode).userObject
        if (model is BaseModel) {
            text = model.name
            icon = ImageIcon(model.icon)
            toolTipText = model.name
        }
        return component
    }
}

private class TreeMouseAdapter(private val treeView: TreeView) : MouseAdapter() {
    override fun mousePressed(event: MouseEvent) {
        if (event.clickCount % 2 == 0) treeView.addTab(event)
    }
}

private class TreeKeyListener(private val treeView: TreeView) : KeyAdapter() {
    override fun keyPressed(keyEvent: KeyEvent) {
        if (keyEvent.extendedKeyCode == KeyEvent.VK_ENTER) treeView.addTab(keyEvent)
    }
}
